File: products/audio-video/libs/common/src/main/java/com/ghatana/audio/video/common/platform/AuthGatewayClient.java
package com.ghatana.audio.video.common.platform;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 * Singleton HTTP client for the platform auth-gateway service.
 *
 * <p>When {@code AUTH_GATEWAY_URL} is set, this client provides a fallback
 * token validation path for platform-issued tokens (e.g. cross-service calls
 * that bypass the Audio-Video local JWT secret). If the environment variable
 * is absent the client returns a disabled singleton that always reports
 * {@link ValidationResult#INVALID}.
 *
 * <p>Usage:
 * <pre>{@code
 *   AuthGatewayClient client = AuthGatewayClient.getInstance();
 *   AuthGatewayClient.ValidationResult result = client.validate(bearerToken);
 *   if (result.valid()) {
 *     String userId = result.userId();   // non-null when valid
 *   }
 * }</pre>
 *
 * @doc.type class
 * @doc.purpose HTTP client for platform auth-gateway token validation
 * @doc.layer platform
 * @doc.pattern Singleton
 */
public final class AuthGatewayClient {

    private static final Logger LOG = LoggerFactory.getLogger(AuthGatewayClient.class);
    private static volatile AuthGatewayClient INSTANCE;

    /**
     * Result of a token validation call to the auth-gateway.
     *
     * @param valid  true if the token is accepted by the platform
     * @param userId authenticated subject, non-null when {@code valid == true}
     * @param email  email associated with the token, may be null
     */
    public record ValidationResult(boolean valid, String userId, String email) {
        /** Stateless invalid singleton to avoid repeated allocation. */
        static final ValidationResult INVALID = new ValidationResult(false, null, null);
    }

    private final String baseUrl;
    private final HttpClient http;
    private final boolean enabled;

    private AuthGatewayClient(String baseUrl) {
        this.baseUrl  = baseUrl;
        this.enabled  = baseUrl != null && !baseUrl.isBlank();
        this.http = enabled
                ? HttpClient.newBuilder()
                        .connectTimeout(Duration.ofSeconds(3))
                        .build()
                : null;

        if (enabled) {
            LOG.info("AuthGatewayClient configured (base={})", baseUrl);
        } else {
            LOG.info("AuthGatewayClient disabled — AUTH_GATEWAY_URL not set");
        }
    }

    /**
     * Returns the process-wide singleton, initialised from {@code AUTH_GATEWAY_URL}.
     *
     * <p>Thread-safe via double-checked locking.
     */
    public static AuthGatewayClient getInstance() {
        if (INSTANCE == null) {
            synchronized (AuthGatewayClient.class) {
                if (INSTANCE == null) {
                    INSTANCE = new AuthGatewayClient(System.getenv("AUTH_GATEWAY_URL"));
                }
            }
        }
        return INSTANCE;
    }

    /** Returns {@code true} when the client is configured and can make real calls. */
    public boolean isEnabled() {
        return enabled;
    }

    /**
     * Validates a raw Bearer token string against the auth-gateway.
     *
     * <p>Sends {@code GET /auth/validate} with the token in an Authorization header
     * and interprets the JSON response. Returns {@link ValidationResult#INVALID}
     * on any network or parse error so callers never see an exception from this
     * method.
     *
     * @param token the raw token (without "Bearer " prefix)
     * @return validation outcome — never {@code null}
     */
    public ValidationResult validate(String token) {
        if (!enabled || token == null || token.isBlank()) {
            return ValidationResult.INVALID;
        }

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + "/auth/validate"))
                    .header("Authorization", "Bearer " + token)
                    .header("Accept", "application/json")
                    .timeout(Duration.ofSeconds(3))
                    .GET()
                    .build();

            HttpResponse<String> response = http.send(request,
                    HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                LOG.debug("AuthGateway rejected token: HTTP {}", response.statusCode());
                return ValidationResult.INVALID;
            }

            return parseValidationResponse(response.body());

        } catch (IOException | InterruptedException e) {
            if (e instanceof InterruptedException) {
                Thread.currentThread().interrupt();
            }
            LOG.warn("AuthGateway call failed: {}", e.getMessage());
            return ValidationResult.INVALID;
        }
    }

    // -------------------------------------------------------------------------
    // Minimal JSON parser — avoids adding a JSON library dependency
    // -------------------------------------------------------------------------

    /**
     * Parses a minimal {@code {valid, userId, email}} JSON payload.
     * Expected format: {@code {"valid":true,"userId":"u1","email":"a@b.com"}}
     */
    private static ValidationResult parseValidationResponse(String json) {
        if (json == null || json.isBlank()) return ValidationResult.INVALID;

        boolean valid  = extractBool(json, "valid");
        String userId  = extractString(json, "userId");
        String email   = extractString(json, "email");

        return valid ? new ValidationResult(true, userId, email) : ValidationResult.INVALID;
    }

    private static boolean extractBool(String json, String key) {
        String pattern = "\"" + key + "\"";
        int idx = json.indexOf(pattern);
        if (idx < 0) return false;
        String rest = json.substring(idx + pattern.length()).stripLeading();
        if (!rest.startsWith(":")) return false;
        rest = rest.substring(1).stripLeading();
        return rest.startsWith("true");
    }

    private static String extractString(String json, String key) {
        String pattern = "\"" + key + "\":\"";
        int start = json.indexOf(pattern);
        if (start < 0) return null;
        start += pattern.length();
        int end = json.indexOf('"', start);
        if (end < 0) return null;
        return json.substring(start, end);
    }
}

File: products/dcmaar/apps/backend/src/services/auth-gateway.client.ts
/**
 * Ghatana Platform Auth-Gateway HTTP Client
 *
 * <p><b>Purpose</b><br>
 * Validates platform-issued JWT tokens against the Ghatana auth-gateway service.
 * Used when DCMAAR needs to accept calls authenticated by the broader Ghatana
 * platform (e.g. from other microservices or the YAPPC frontend).
 *
 * <p><b>Design</b><br>
 * The client is intentionally optional — when {@code AUTH_GATEWAY_URL} is not
 * configured, all validation calls return an unauthenticated identity so that
 * DCMAAR can continue to run standalone with its own local JWT auth.
 *
 * <p><b>API contract</b><br>
 * {@code GET /auth/validate} with {@code Authorization: Bearer <token>}
 * → {@code { valid: boolean, userId?: string, email?: string, error?: string }}
 *
 * <p><b>Usage</b><br>
 * <pre>{@code
 * const identity = await AuthGatewayClient.getInstance().validate(token);
 * if (identity.valid) {
 *   request.platformUserId = identity.userId;
 * }
 * }</pre>
 *
 * @doc.type class
 * @doc.purpose Platform auth-gateway token validation client
 * @doc.layer backend
 * @doc.pattern Adapter
 */
import { logger } from '../utils/logger';

const VALIDATE_PATH = '/auth/validate';
const DEFAULT_TIMEOUT_MS = 5_000;

/** Identity returned by the auth-gateway validate endpoint. */
export interface PlatformIdentity {
  valid: boolean;
  userId?: string;
  email?: string;
  /** Roles assigned to this identity by the platform IAM system. */
  roles?: readonly string[];
  error?: string;
}

/** Raw auth-gateway validate response. */
interface AuthGatewayValidateResponse {
  valid?: boolean;
  userId?: string;
  email?: string;
  roles?: string[];
  error?: string;
}

export class AuthGatewayClient {
  private static instance: AuthGatewayClient;

  private readonly baseUrl: string | null;

  private constructor() {
    this.baseUrl = process.env['AUTH_GATEWAY_URL'] ?? null;

    if (this.baseUrl) {
      logger.info('AuthGatewayClient initialised', { baseUrl: this.baseUrl });
    } else {
      logger.warn(
        'AUTH_GATEWAY_URL not configured — platform token validation disabled. ' +
          'DCMAAR will operate in standalone auth mode.',
      );
    }
  }

  /**
   * Returns the singleton instance. Thread-safe via module-level singleton.
   */
  public static getInstance(): AuthGatewayClient {
    if (!AuthGatewayClient.instance) {
      AuthGatewayClient.instance = new AuthGatewayClient();
    }
    return AuthGatewayClient.instance;
  }

  /**
   * Validates a platform-issued Bearer token against the auth-gateway.
   *
   * @param token  raw JWT token string (without the "Bearer " prefix)
   * @returns resolved identity if valid; {@code { valid: false }} when the
   *          gateway is unreachable or the token is rejected
   */
  async validate(token: string): Promise<PlatformIdentity> {
    if (!this.baseUrl) {
      // Gateway not configured — cannot validate platform tokens in this deployment
      return { valid: false, error: 'AUTH_GATEWAY_URL not configured' };
    }

    const url = `${this.baseUrl}${VALIDATE_PATH}`;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), DEFAULT_TIMEOUT_MS);

    try {
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          Authorization: `Bearer ${token}`,
          Accept: 'application/json',
        },
        signal: controller.signal,
      });

      const body = (await response.json()) as AuthGatewayValidateResponse;

      if (!response.ok || body.valid === false) {
        logger.debug('AuthGateway rejected token', { status: response.status, error: body.error });
        return { valid: false, error: body.error ?? 'Token rejected by auth-gateway' };
      }

      return {
        valid: true,
        userId: body.userId,
        email: body.email,
        roles: body.roles,
      };
    } catch (err) {
      // Network error or timeout — fail open to avoid blocking DCMAAR's own auth path
      logger.warn('AuthGateway validation request failed — treating token as invalid', {
        url,
        error: err instanceof Error ? err.message : String(err),
      });
      return { valid: false, error: 'Auth-gateway unreachable' };
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Returns {@code true} when the auth-gateway URL is configured and the service
   * can be used for platform token validation.
   */
  isConfigured(): boolean {
    return this.baseUrl !== null;
  }
}

export const authGatewayClient = AuthGatewayClient.getInstance();
